### ISSUE TITLE
use poll reactor on POSIX platform that provide poll(2)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2234]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2234|
|Type|enhancement|
|Created|2006-11-15 15:50:24Z|
|Branch|https://github.com/twisted/twisted/tree/poll-default-2234-3|

Currently the default is always select.  We can do better than this.

<details><summary>Searchable metadata</summary>

```
trac-id__2234 2234
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_poll_default_2234_3 branches/poll-default-2234-3
branch_author__itamarst__exarkun itamarst, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1163605824000000 1163605824000000
changetime__1304563687000000 1304563687000000
version__None None
owner__itamar itamar
cc__zooko cc__zooko@... cc__itamar
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_2"></a>|@zooko commented|
|-|-|

Tahoe, the Least-Authority Filesystem, currently has the following reactor-selection login in its Makefile:

http://allmydata.org/trac/tahoe/browser/Makefile?rev=2847

```
17	PLAT = $(strip $(shell $(PYTHON) -c "import sys ; print sys.platform"))
...
25	        ifeq ($(PLAT),linux2)
26	                # This is to work-around #402, and anyway the poll reactor is probably better on Linux, if
27	                # we have a lot of open fds.
28	                ifeq ($(REACTOR),)
29	                        REACTOR := poll
30	                endif
31	        endif
32	        ifeq ($(PLAT),cygwin)
33	                # The cygwin select reactor seems to run out of fds in unit tests -- it writes "filedescriptor
34	                # out of range in select()".  Setting reactor=poll fixes that.
35	                ifeq ($(REACTOR),)
36	                        REACTOR := poll
37	                endif
38	        endif
... 
46	ifneq ($(REACTOR),)
47	        REACTOROPT := --reactor=$(REACTOR)
48	else
49	        REACTOROPT :=
50	endif
```

It seems like poll reactor could be the default reactor on platforms which support poll reactor, and then Tahoe could stop doing anything about reactor choice at all in its Makefile.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_3"></a>|@zooko commented|
|-|-|

The following tahoe issue ticket now contains a link to this ticket:

http://allmydata.org/trac/tahoe/ticket/402 # bug in Twisted, triggered by pyOpenSSL-0.7

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

You lost me at "Makefile".  Would you mind rephrasing that chunk of stuff in a language that humans understand - like, say, Python? :-)

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_5"></a>|@warner commented|
|-|-|

As background, the Makefile stuff that Zooko quoted above is from the Tahoe
project's top-level Makefile, which really just contains convenient aliases
so I can type 'make test' instead of 'PYTHONPATH=localstuff trial
--reactor=poll allmydata.test'.

To work around [#3218](https://github.com/twisted/twisted/issues/3218) (which causes foolscap and tahoe test failures under
Twisted-8.x and pyOpenSSL-0.7 on the select reactor), we changed this
Makefile to add --reactor=poll to the trial command line, but since not all
platforms support pollreactor, there are a bunch of conditionals to make sure
we don't use --reactor=poll on, say, windows. (we might have something
elsewhere to use iocpreactor there).

If trial could be conveniently invoked from python code, we would probably
replace this 'make test' target with a python script that did something like:

```
from twisted.python import runtime
r = "default"
if runtime.platformType == "posix":
    r = "poll"
sys.exit(trialRunner(suite, reactor=r))
```

(in fact, zooko would probably write a setuptools plugin to make 'python
setup.py test' do exactly this :-).

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_6"></a>|@zooko commented|
|-|-|

In fact, I've already written such a setuptools plugin, by borrowing code from the Elise project, but there may have been some problem using it, or else I just got distracted before I finished it.  I'll probably dig it up again, especially since Barry Warsaw just wished for better setuptools integration for tests: http://mail.python.org/pipermail/python-dev/2008-August/081764.html

But, actually this has little to do with the topic of this ticket, which is that choosing the select reactor as the default doesn't seem as good as choosing poll reactor as the default, if there is a poll reactor available.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Poll shouldn't be used by default on OS X.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_10"></a>|@zooko commented|
|-|-|

Okay, so how about if the default reactor is poll on linux and cygwin and select on everything else?  That's what the Tahoe makefile currently does.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

cygwin isn't an officially supported platform.

Other than that, no problems with that suggestion are immediately obvious.

Do you want to supply a patch?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_12"></a>|@zooko set owner to @zooko<br>@zooko set status to `assigned`|
|-|-|

I would be interested in contributing a cygwin buildslave.

Yes, I want to supply a patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Cool, a cygwin slave would be great.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_14"></a>|@zooko commented|
|-|-|

Please provide name, password, and master for the cygwin buildbot.  Also, would you like a cygwin buildbot for pyOpenSSL too?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

A comment on the tracker wouldn't be a good place for most of that information.  Find me on IRC.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_16"></a>|@zooko commented|
|-|-|

Okay, we have a cygwin buildslave now, but it goes into a fast loop and fills my filesystem with test log:

http://buildbot.twistedmatrix.com/builders/cygwin-py2.5-select/builds/3

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_17"></a>|@zooko commented|
|-|-|

[#3529](https://github.com/twisted/twisted/issues/3529) is a bug which prevents me from running a buildslave on cygwin.  Until that bug is fixed or worked-around in some way so that I can run a cygwin buildslave, I will delay working on this issue.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_18"></a>|@zooko commented|
|-|-|

This issue also affects ```setuptools_trial``` -- [setuptools_trial issue #2](http://allmydata.org/trac/setuptools_trial/ticket/2)

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_19"></a>|@zooko commented|
|-|-|

This ticket has been mentioned on the tahoe-dev list:
http://allmydata.org/pipermail/tahoe-dev/2009-January/001027.html

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_20"></a>|@zooko commented|
|-|-|

So I was just hacking setuptools_trial to select the poll reactor on cygwin or linux2:

http://allmydata.org/trac/setuptools_trial/changeset/13

When I realized -- hey wait a minute, why am I waiting for the cygwin buildslave to work?  Why don't we fix this ticket for all the platforms which are already passing Twisted buildbot tests and worry about cygwin once we have a working cygwin buildslave?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

[#4153](https://github.com/twisted/twisted/issues/4153) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

I agree.  How about restricting the scope of this ticket.  I propose just one thing: if the platform is linux, use poll reactor as the default; for all other cases, continue to use select reactor as the default.

Once this is implemented, we can file tickets for further refinements to the selection scheme.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_23"></a>|@zooko commented|
|-|-|

Good idea!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [28660]) Branching to 'poll-default-2234'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [28661]) Parameterize more of how Platform works so I can pretend to be on OS X when I'm on Linux

refs [#2234](https://github.com/twisted/twisted/issues/2234)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

There's a stdio/pollreactor issue.  I filed [#4352](https://github.com/twisted/twisted/issues/4352) for it.  This issue is blocked until that's resolved.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_27"></a>|@itamarst commented|
|-|-|

As long as we're changing this, I'd want these defaults:

Linux: epoll, switching to POSIX default if unavailable
POSIX: poll if available, otherwise select

Simply trying epoll,poll,select in that order on POSIX platforms would accomplish this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

Eventually, yes.  I'm not going to do all that, though.  I'm just going to make a simple change.  Further customizing the selection logic should be easier based on the outcome of this ticket, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [28680]) Branching to 'poll-default-2234-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst commented|
|-|-|

A note on the current implementation: many platforms are POSIX but not Mac OS X, not just Linux. One *hopes* all of them have poll, but somehow that seems too much to ask... It would be nice not to break the default reactor on such platforms, even if they're unsupported.

Checking for existence of poll object in select module might be better.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_31"></a>|@itamarst commented|
|-|-|

(In [31590]) Branching to 'poll-default-2234-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_32"></a>|@itamarst removed owner<br>@itamarst set status to `new`|
|-|-|

Ready for review, I think. Buildbot run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/poll-default-2234-3

The problems appear to be normal ongoing windows issues.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_33"></a>|@zooko commented|
|-|-|

Reviewed: I read through this patch and didn't see anything wrong with it. I liked the clarification that this isn't actually choosing the "best" reactor.

(Do I need to do anything else to mark this ticket as reviewed?)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_34"></a>|@itamarst commented|
|-|-|

Standard procedure is to remove "review" keyword and reassign back to author.

ReviewProcess documents the things you should be checking.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)|@zooko set owner to @zooko<br>@zooko set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [31595]) Remove copyright date

refs [#2234](https://github.com/twisted/twisted/issues/2234)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

What's this stuff about "modern OS X"?  OS X 10.6.7, the Apple distributed Python executable does not have it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_38"></a>|@itamarst commented|
|-|-|

AFAICT the *OS* at least has it, which is why I assumed Python exposed it:
http://developer.apple.com/library/mac/#documentation/Darwin/Reference/ManPages/man2/poll.2.html

According to this - http://bugs.python.org/issue5154 (as usual, foom knows the answer to everything), Apple chose not to expose because certain fds aren't supported. The only relevant one is PTYs, but that is an issue (why does OS X suck?).

The implication of the linked bug is that some releases of Python might decide expose a still-broken poll, so possibly hardcoding select() on OS X is still the right thing to do... Maybe we should ask James, since he's well informed.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_39"></a>|@zooko set owner to @itamarst<br>@zooko set status to `new`|
|-|-|

Okay, I have reviewed that patch and checked for the things mentioned in ReviewProcess. Other than the issue raised in comment:38, it is ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_40"></a>|@itamarst removed owner|
|-|-|

I reverted back to old behavior (no poll on OSX), and added description of why we are not choosing e.g. epoll on Linux just yet.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/poll-default-2234-3 doesn't show any relevant errors AFAICT.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks!

  1. Please remove the date from `test_default.py`'s copyright header
  1. Can you change the docstring style for `isMacOSX` in `runtime.py` too?
  1. I think the news fragment would benefit from a "The ..." at the beginning.

I wonder if the level of detail now in the default reactor's description (in the `twisted_reactors.py` dropin file) will scale to more complex selection schemes.  We'll find out, I guess. :)

Please address the above to your satisfaction and then merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_42"></a>|@glyph commented|
|-|-|

I just noticed http://twistedmatrix.com/trac/changeset/31761 go by in my RSS reader, and I have one comment:

We shouldn't have FUD about our own code, especially in comments.  The issues with epoll on Linux and poll(2) on OS X are nicely well-described.  But the issues with kqueue ("isn't in great shape") and IOCP ("probably not quite ready for prime time") are super vague, and, in the latter case, I think possibly wrong.

Please link to specific issues in that comment so that a future maintainer can know whether it's an appropriate time to update the code to select a better default reactor.  (Also, while you're at it, transforming "[#4429](https://github.com/twisted/twisted/issues/4429)" into the appropriate URL would be a kindness to future readers.)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_43"></a>|@itamarst commented|
|-|-|

Replying to [glyph](#note_42):

I'll make the descriptions more accurate (remind me what the kqueue issues are? besides PTY support.)

That being said: if IOCP really doesn't have any reported issues (which I doubt, but will verify) would you be willing to have IOCP the default on Windows?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_44"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_43):
> Replying to [glyph](#note_42):
> 
> I'll make the descriptions more accurate (remind me what the kqueue issues are? besides PTY support.)

PTY support's a big one, but ... hrm.  [ticket:1918 the relevant ticket] is a big mess that doesn't really describe the issues, but that's probably the thing to link to.  Somebody should update the summary there.  Possibly in the comment you could just say that it is "out of date".

> That being said: if IOCP really doesn't have any reported issues (which I doubt, but will verify) would you be willing to have IOCP the default on Windows?

"any" reported issues? :)  I think the important thing would be _fewer_ reported issues than the select reactor on Windows :).  Well, that, and no significant missing functionality.  SSL was the big one for a long time, but now that that's taken care of, what's left?   Do serial ports work?  Standard I/O?  I know that subprocesses use the same gross hack as all the other Windows reactors.

I'd definitely be willing - happy, even - to have it be the default, but I'm happy to be overridden here by anyone with an issue that I haven't thought of (Pavel or Jean-Paul might know about something).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_45"></a>|@itamarst set status to `closed`|
|-|-|

(In [31768]) Merge poll-default-2234-3.

Author: exarkun, itamar
Reviewer: exarkun, glyph
Fixes: [#2234](https://github.com/twisted/twisted/issues/2234)

On platforms that support it (Linux, FreeBSD, and other non-OS X POSIX platforms) use poll reactor by default.
